Refactor BucketBatch to BucketBatcher

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#63185 Refactor BucketBatch**
* #63184 Replace str by repr for DataChunk


- To land BucketBatch in `torchdata`, functional API needs to be removed.
- Refactor the BucketBatch to be same as `torchdata`
- Add deprecation warning to users
- Add corresponding tests

Differential Revision: [D30288893](https://our.internmc.facebook.com/intern/diff/D30288893)

ejguan committed:
Refactor BucketBatch

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30288893).

ejguan committed:
Update on "Refactor BucketBatch"


- To land BucketBatch in `torchdata`, functional API needs to be removed.
- Refactor the BucketBatch to be same as `torchdata`
- Add deprecation warning to users
- Add corresponding tests

Differential Revision: [D30288893](https://our.internmc.facebook.com/intern/diff/D30288893)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30288893).

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -711,52 +711,48 @@ def test_unbatch_datapipe(self):
+
+        # In-batch shuffle
+        _helper(batch_size=3, drop_last=False, batch_num=5, sort_key=_sort_fn)
+        #  _helper(batch_size=3, drop_last=False, batch_num=2, bucket_num=2, sort_key=_sort_fn)
```
nit: Drop comments?

VitalyFedyunin commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -132,9 +133,13 @@ def _dive(self, element, unbatch_level):
             else:
                 raise IndexError(f"unbatch_level {self.unbatch_level} exceeds the depth of the DataPipe")
 
+# TODO(Erjia): https://github.com/pytorch/pytorch/issues/63095
```
super nit: better TODO(accout)

ejguan committed:
Update on "Refactor BucketBatch"



- To land BucketBatch in `torchdata`, functional API needs to be removed.
- Refactor the BucketBatch to be same as `torchdata`
- Add deprecation warning to users
- Add corresponding tests

Differential Revision: [D30288893](https://our.internmc.facebook.com/intern/diff/D30288893)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30288893).
